src/Compression/CompressionCodecEncrypted.cpp

```diff
@@ -11,204 +19,291 @@
+{
+    if (!params.get())
+        throw Exception("Empty params in CompressionCodecEncrypted configuration", ErrorCodes::BAD_ARGUMENTS);
+    return params.get()->current_key_id;
+}
+
+String CompressionCodecEncrypted::Configuration::getNonce() const
+{
+    if (!params.get())
+        throw Exception("Empty params in CompressionCodecEncrypted configuration", ErrorCodes::BAD_ARGUMENTS);
+    return params.get()->nonce;
+}
+
+CompressionCodecEncrypted::CompressionCodecEncrypted()
+{
+    setCodecDescription("AES_128_GCM_SIV");
+}
+
+uint8_t CompressionCodecEncrypted::getMethodByte() const
+{
+    return static_cast<uint8_t>(CompressionMethodByte::Encrypted);
+}
+
+void CompressionCodecEncrypted::updateHash(SipHash & hash) const
+{
+    getCodecDesc()->updateTreeHash(hash);
+}
+
+UInt32 CompressionCodecEncrypted::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    // The GCM mode is a stream cipher. No paddings are
+    // involved. There will be a tag at the end of ciphertext (16
+    // octets).
+    return uncompressed_size + 24;
+}
+
+UInt32 CompressionCodecEncrypted::doCompressData(const char * source, UInt32 source_size, char * dest) const
+{
+    // Generate an IV out of the data block and the key-generation
+    // key. It is completely deterministic, but does not leak any
+    // information about the data block except for equivalence of
+    // identical blocks (under the same master key). The IV will
+    // be used as an authentication tag. The ciphertext and the
+    // tag will be written directly in the dest buffer.
+    const std::string_view plaintext = std::string_view(source, source_size);
+
+    encrypt(plaintext, dest);
+    return source_size + 24;
+}
+
+void CompressionCodecEncrypted::doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 uncompressed_size [[maybe_unused]]) const
+{
+    UInt64 encrypted_text_key_id;
+    source = readVarUInt(encrypted_text_key_id, source, 8);
+    source_size -= length_of_varint;
+    // Extract the IV from the encrypted data block. Decrypt the
+    // block with the extracted IV, and compare the tag. Throw an
+    // exception if tags don't match.
+    const std::string_view ciphertext_and_tag = std::string_view(source, source_size);
+    assert(ciphertext_and_tag.size() == uncompressed_size + 16);
+
+    decrypt(ciphertext_and_tag, dest, encrypted_text_key_id);
+}
+
+std::string CompressionCodecEncrypted::lastErrorString()
+{
+    std::array<char, 1024> buffer;
+    ERR_error_string_n(ERR_get_error(), buffer.data(), buffer.size());
+    return std::string(buffer.data());
+}
+
+void CompressionCodecEncrypted::encrypt(const std::string_view & plaintext, char * ciphertext_and_tag)
+{
+    auto& conf = Configuration::instance();
+    String nonce_from_config = conf.getNonce();
+    std::string_view nonce("\0\0\0\0\0\0\0\0\0\0\0\0", 12);
+    if (!nonce_from_config.empty())
+        nonce = nonce_from_config;
```

Comment:
Please move this logic to the `Configuration` class. I mean you could make `getNonce()` to return zero nonce instead of an empty one.